org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java

```diff
@@ -2057,6 +2060,10 @@ private String getAnnotationType(IJavaProject javaProject, List<String> annotati
 				if (classpathStorage.keySet().contains(annotationType)) {
 					// for known types, check the classpath to achieve a better performance
 					for (String classpath : result.classpaths) {
+						IClasspathEntry classpathEntry = javaProject.getClasspathEntryFor(new Path(classpath));
+						if (isTest(classpathEntry, javaProject)) {
```

Comment:
how about directly use `ClasspathOptions.scope = "runtime"` in https://github.com/eclipse/eclipse.jdt.ls/blob/f8452b422add64f9bbc80d0e22b48ed71797e563/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java#L2055

Reply:
Agree. We could exclude test scope directly when getting the project classpaths.

Reply:
it seems that using `ClasspathOptions.scope = "runtime"` doesn't work in this case. I just tested sample project in https://github.com/redhat-developer/vscode-java/issues/2712#issuecomment-1269828190 and see the introduced artifact jsr 305 is in runtime scope:

![image](https://user-images.githubusercontent.com/45906942/194691229-e35da2f7-d20a-4b26-a033-fea6a83f0d48.png)

So the scope filter won't filter this dependency. However, in the classpath entry, it has test attribute:

![image](https://user-images.githubusercontent.com/45906942/194691290-659602c0-3286-41e9-ada8-95b2ded2a54f.png)

So the new introduced method `isTest()` works. it will check the attributes of the classpath entry.

Reply:
and I also tried `ProjectCommand.isTestClasspathEntry()` mentioned by @jdneo , it also checks the attribute and it works well. So it seems that we don't need to add a new method.

Reply:
> it seems that using ClasspathOptions.scope = "runtime" doesn't work in this case. I just tested sample project in https://github.com/redhat-developer/vscode-java/issues/2712#issuecomment-1269828190 and see the introduced artifact jsr 305 is in runtime scope:

This is a bug on the utility method `ProjectCommand.getClasspathsFromJavaProject`, which never uses the passed `scope` parameter.
https://github.com/eclipse/eclipse.jdt.ls/blob/f8452b422add64f9bbc80d0e22b48ed71797e563/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/JavaLaunchConfigurationInfo.java#L35-L50

In the launch configuration template `JAVA_APPLICATION_LAUNCH`, we could use the following attribute to exclude test code.
`<booleanAttribute key="org.eclipse.jdt.launching.ATTR_EXCLUDE_TEST_CODE" value="true"/>`

Reply:
It's interesting that the test attribute is set to `true` but the `maven.scope` is `runtime`.

Is that a bug of m2e?

Reply:
> which never uses the passed scope parameter.

It will,

https://github.com/eclipse/eclipse.jdt.ls/blob/f8452b422add64f9bbc80d0e22b48ed71797e563/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/JavaApplicationLaunchConfiguration.java#L43-L48

Reply:
FYI the jsr305 dependency is declared as `implementation` scope, not `test` in spring-kafka: https://github.com/spring-projects/spring-kafka/blob/4bcb503a2fb33d20b45f4cc241c769c397bf0a7d/build.gradle#L173